Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -721,13 +709,66 @@ namespace Babylon
+    }
+
+    Napi::Value NativeEngine::CreateProgram(const Napi::CallbackInfo& info)
+    {
+        const std::string vertexSource = info[0].As<Napi::String>().Utf8Value();
+        const std::string fragmentSource = info[1].As<Napi::String>().Utf8Value();
+        ProgramData* program = new ProgramData{};
+        try
+        {
+            std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>> programInfo = CreateProgramInternal(vertexSource, fragmentSource, *program);
+            program->Handle = programInfo.first;
+            program->VertexAttributeLocations = std::move(programInfo.second);
+        }
+        catch (const std::exception& ex)
+        {
+            delete program;
+            throw Napi::Error::New(info.Env(), ex.what());
+        }
+        return Napi::Pointer<ProgramData>::Create(info.Env(), program, Napi::NapiPointerDeleter(program));
+    }
 
-        program->Handle = bgfx::createProgram(vertexShader, fragmentShader, true);
+    Napi::Value NativeEngine::CreateProgramAsync(const Napi::CallbackInfo& info)
+    {
+        const std::string vertexSource = info[0].As<Napi::String>().Utf8Value();
+        const std::string fragmentSource = info[1].As<Napi::String>().Utf8Value();
+        const Napi::Function onSuccess = info[2].As<Napi::Function>();
+        const Napi::Function onError = info[3].As<Napi::Function>();
+        ProgramData* program = new ProgramData{};
+        arcana::make_task(arcana::threadpool_scheduler, *m_cancellationSource,
+            [this, vertexSource, fragmentSource, program, cancellationSource{m_cancellationSource}]() -> std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>>
+            {
+                return CreateProgramInternal(vertexSource, fragmentSource, *program);
+            })
+            .then(m_runtimeScheduler, *m_cancellationSource, [program, onSuccessRef{Napi::Persistent(onSuccess)}, onErrorRef{Napi::Persistent(onError)}, cancellationSource{m_cancellationSource}](arcana::expected<std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>>, std::exception_ptr> result)
+                {
+                    if (result.has_error())
+                    {
+                        try
+                        {
+                            std::rethrow_exception(result.error());
+                        }
+                        catch (const std::exception& ex)
+                        {
+                            onErrorRef.Call({Napi::Error::New(onErrorRef.Env(), ex.what()).Value()});
+                        }
+                    }
+                    else
+                    {
+                        std::pair<bgfx::ProgramHandle, std::unordered_map<std::string, uint32_t>> programInfo = result.value();
+                        program->Handle = programInfo.first;
+                        program->VertexAttributeLocations = std::move(programInfo.second);
```

Comment:
@bghgary, I don't think you mentioned this one, but I'm a little worried, is it dangerous to do this with the `program` pointer? Given that this can happen after `CreateProgramAsync` returns, what happens if the `NapiPointerDeleter` deletes it before this runs? (Sorry if I'm misunderstanding the functionality of these APIs.)

Reply:
This is a good callout. Without the correct pattern for shutdown, it can indeed get called after the JS object is deleted which in turn deletes the program. If you want to find out more about how the new pattern works, see [this PR](https://github.com/BabylonJS/BabylonNative/pull/1141), specifically [this comment](https://github.com/BabylonJS/BabylonNative/pull/1141/files#diff-a4dbd5b4fd890908e7876413e36816689a92f3a30cd8260147d840604dba29b5R10-R55).

Reply:
Also, note that NapiPointerDeleter only deletes the object when garbage collected. It will not be deleted until the JS side lets go of the JS object, which shouldn't happen until `DeleteProgram` is called. This NapiPointerDeleter is a safe guard in case the JS engine goes down before the engine is disposed.

Reply:
I briefly read the file you linked to above, but it's not clear to me how it applies to what I'm doing here. What I'm doing still feels unsafe (regardless of the behavior of the BGFX resource, just dereferencing the deleted pointer at all sounds bad), unless you're saying I can rely on the `NapiPointerDeleter` not to run on this pointer until the Javascript environment shuts down, but that's actually not something I would want to rely on even if I could, because of memory leaks.

Just now I looked at the actual code behind `NapiPointerDeleter` and it's simpler than I thought. Would it be okay to replace that with a different callback that only conditionally deletes the pointer? I might be able to make it safe that way, but it sounds a bit convoluted and maybe I'm missing some API already in here that is meant for this sort of thing.

Reply:
> unless you're saying I can rely on the NapiPointerDeleter not to run on this pointer until the Javascript environment shuts down

You can rely on the JS side holding onto the resulting JS object until `DeleteProgram` is called. If `DeleteProgram` doesn't get called, then the NapiPointerDeleter is here to make sure it is still deleted.

Reply:
That makes sense, but I'm concerned about what happens if it goes the other way around: While the shader is compiling (which might take a few frames) the Javascript side stops using that object and decides to GC it, `NapiPointerDeleter` gets called (or at least I assume that's what it's there for), and *then* the C++ side comes back with a successful compilation and tries to assign `program->Handle` etc onto (I assume) a pointer to invalid memory.

Reply:
Right, the coding pattern I mentioned above will prevent the JavaScript object from being destroyed because the continuation will capture a strong ref to the object. This can't be done until my PR is merged, so you don't have to worry about it here.

Reply:
I've pushed the changes we discussed, please let me know if there's anything wrong with the new version.